File: src/modules/request/services/request.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { DataSource, Repository } from 'typeorm';
import { AccessControlService } from '../../../common/services/access-control.service';
import { AuthenticatedUser } from '../../../common/types/authenticated-user.type';
import { ApprovalStep } from '../../approval/entities/approval-step.entity';
import { ApprovalRole } from '../../approval/enums/approval-role.enum';
import { ApprovalStepStatus } from '../../approval/enums/approval-step-status.enum';
import { Employee } from '../../employee/entities/employee.entity';
import { EquipmentCategory } from '../../equipment-category/entities/equipment-category.entity';
import { EquipmentModel } from '../../equipment-model/entities/equipment-model.entity';
import { NotificationService } from '../../notification/services/notification.service';
import { CancelRequestDto } from '../dto/cancel-request.dto';
import { CreateRequestDto } from '../dto/create-request.dto';
import { EquipmentRequest } from '../entities/equipment-request.entity';
import {
  CANCELLABLE_REQUEST_STATUSES,
  RequestStatus,
} from '../enums/request-status.enum';
import { RequestType } from '../enums/request-type.enum';

@Injectable()
export class RequestService {
  constructor(
    @InjectRepository(EquipmentRequest)
    private readonly requestRepository: Repository<EquipmentRequest>,
    @InjectRepository(EquipmentModel)
    private readonly modelRepository: Repository<EquipmentModel>,
    @InjectRepository(EquipmentCategory)
    private readonly categoryRepository: Repository<EquipmentCategory>,
    @InjectRepository(Employee)
    private readonly employeeRepository: Repository<Employee>,
    private readonly dataSource: DataSource,
    private readonly notificationService: NotificationService,
    private readonly accessControl: AccessControlService,
  ) {}

  findMyRequests(userId: string): Promise<EquipmentRequest[]> {
    return this.requestRepository.find({
      where: { requester: { id: userId } },
      relations: {
        requester: true,
        equipmentModel: { category: true },
        category: true,
        approvalSteps: { approver: true },
        alternatives: { equipmentModel: true, suggestedBy: true },
      },
      order: { createdAt: 'DESC', approvalSteps: { level: 'ASC' } },
    });
  }

  async findOne(
    id: string,
    user: AuthenticatedUser,
  ): Promise<EquipmentRequest> {
    const request = await this.requestRepository.findOne({
      where: { id },
      relations: {
        requester: { department: true },
        equipmentModel: { category: true },
        category: true,
        approvalSteps: { approver: true },
        alternatives: { equipmentModel: true, suggestedBy: true },
        assignments: { equipmentAsset: { equipmentModel: true } },
      },
      order: { approvalSteps: { level: 'ASC' } },
    });

    if (!request) {
      throw new NotFoundException(`Request with id "${id}" not found`);
    }

    await this.assertCanViewRequest(request, user);
    return request;
  }

  async getTimeline(id: string, user: AuthenticatedUser) {
    const request = await this.findOne(id, user);
    return {
      requestId: request.id,
      status: request.status,
      steps: (request.approvalSteps ?? []).map((step) => ({
        id: step.id,
        level: step.level,
        approverRole: step.approverRole,
        approverName: `${step.approver.firstName} ${step.approver.lastName}`,
        status: step.status,
        comment: step.comment,
        actedAt: step.actedAt,
      })),
      alternatives: request.alternatives ?? [],
      assignments: request.assignments ?? [],
    };
  }

  async create(
    createRequestDto: CreateRequestDto,
    user: AuthenticatedUser,
  ): Promise<EquipmentRequest> {
    const requester = await this.employeeRepository.findOne({
      where: { id: user.id },
      relations: { department: { directManager: true } },
    });

    if (!requester) {
      throw new NotFoundException('Requester not found');
    }

    if (!requester.department?.directManager) {
      throw new BadRequestException(
        'Cannot submit request: department has no direct manager assigned',
      );
    }

    let equipmentModel: EquipmentModel | undefined;
    let category: EquipmentCategory | undefined;

    if (createRequestDto.requestType === RequestType.LOAN) {
      if (!createRequestDto.equipmentModelId) {
        throw new BadRequestException(
          'equipmentModelId is required for loan requests',
        );
      }
      equipmentModel =
        (await this.modelRepository.findOne({
          where: { id: createRequestDto.equipmentModelId },
          relations: { category: true },
        })) ?? undefined;
      if (!equipmentModel) {
        throw new NotFoundException('Equipment model not found');
      }
      category = equipmentModel.category;
    } else {
      if (!createRequestDto.requestedItemName || !createRequestDto.categoryId) {
        throw new BadRequestException(
          'requestedItemName and categoryId are required for procurement requests',
        );
      }
      category =
        (await this.categoryRepository.findOne({
          where: { id: createRequestDto.categoryId },
        })) ?? undefined;
      if (!category) {
        throw new NotFoundException('Category not found');
      }
    }

    const savedRequest = await this.dataSource.transaction(async (manager) => {
      const requestRepo = manager.getRepository(EquipmentRequest);
      const stepRepo = manager.getRepository(ApprovalStep);

      const request = requestRepo.create({
        requester,
        requestType: createRequestDto.requestType,
        equipmentModel,
        requestedItemName: createRequestDto.requestedItemName,
        category,
        quantity: createRequestDto.quantity ?? 1,
        startDate: createRequestDto.startDate,
        endDate: createRequestDto.endDate,
        purpose: createRequestDto.purpose,
        status: RequestStatus.PENDING_MANAGER_APPROVAL,
      });

      const persisted = await requestRepo.save(request);

      const managerStep = stepRepo.create({
        request: persisted,
        level: 1,
        approver: requester.department!.directManager!,
        approverRole: ApprovalRole.DIRECT_MANAGER,
        status: ApprovalStepStatus.PENDING,
      });
      await stepRepo.save(managerStep);

      return requestRepo.findOneOrFail({
        where: { id: persisted.id },
        relations: {
          requester: true,
          equipmentModel: { category: true },
          category: true,
          approvalSteps: { approver: true },
        },
        order: { approvalSteps: { level: 'ASC' } },
      });
    });

    const firstStep = savedRequest.approvalSteps?.[0];
    if (firstStep?.approver) {
      await this.notificationService.notifyApprovalRequired(
        firstStep.approver,
        savedRequest,
        firstStep,
      );
    }

    return savedRequest;
  }

  async cancel(
    id: string,
    user: AuthenticatedUser,
    cancelDto: CancelRequestDto,
  ): Promise<EquipmentRequest> {
    const request = await this.findOne(id, user);

    if (request.requester.id !== user.id) {
      throw new ForbiddenException('You can only cancel your own requests');
    }

    if (!CANCELLABLE_REQUEST_STATUSES.includes(request.status)) {
      throw new BadRequestException(
        'Request can only be cancelled while pending manager or procurement approval',
      );
    }

    request.status = RequestStatus.CANCELLED;
    request.cancellationReason = cancelDto.reason;
    request.cancelledAt = new Date();

    const saved = await this.requestRepository.save(request);
    await this.notificationService.notifyRequestCancelled(
      request.requester,
      saved,
    );
    return saved;
  }

  private async assertCanViewRequest(
    request: EquipmentRequest,
    user: AuthenticatedUser,
  ): Promise<void> {
    if (request.requester.id === user.id) {
      return;
    }

    if (this.accessControl.isAdmin(user)) {
      return;
    }

    if (this.accessControl.isProcurementManagerOrAbove(user)) {
      return;
    }

    if (this.accessControl.isDirectManagerOrAbove(user)) {
      const canView = await this.accessControl.isEmployeeInManagedDepartments(
        user.id,
        request.requester.id,
      );
      if (canView) return;
    }

    throw new ForbiddenException('You cannot view this request');
  }

  async findManagerPending(
    user: AuthenticatedUser,
  ): Promise<EquipmentRequest[]> {
    const departmentIds = await this.accessControl.getManagedDepartmentIds(
      user.id,
    );
    if (!departmentIds.length) {
      return [];
    }

    const employeeIds =
      await this.accessControl.getDepartmentEmployeeIds(departmentIds);

    return this.requestRepository
      .createQueryBuilder('request')
      .leftJoinAndSelect('request.requester', 'requester')
      .leftJoinAndSelect('request.equipmentModel', 'equipmentModel')
      .leftJoinAndSelect('request.category', 'category')
      .leftJoinAndSelect('request.approvalSteps', 'approvalSteps')
      .leftJoinAndSelect('approvalSteps.approver', 'approver')
      .where('request.status = :status', {
        status: RequestStatus.PENDING_MANAGER_APPROVAL,
      })
      .andWhere('requester.id IN (:...employeeIds)', { employeeIds })
      .orderBy('request.createdAt', 'DESC')
      .getMany();
  }

  async findManagerRequests(
    user: AuthenticatedUser,
  ): Promise<EquipmentRequest[]> {
    const departmentIds = await this.accessControl.getManagedDepartmentIds(
      user.id,
    );
    if (!departmentIds.length) {
      return [];
    }

    const employeeIds =
      await this.accessControl.getDepartmentEmployeeIds(departmentIds);

    return this.requestRepository
      .createQueryBuilder('request')
      .leftJoinAndSelect('request.requester', 'requester')
      .leftJoinAndSelect('request.equipmentModel', 'equipmentModel')
      .leftJoinAndSelect('request.category', 'category')
      .leftJoinAndSelect('request.approvalSteps', 'approvalSteps')
      .leftJoinAndSelect('approvalSteps.approver', 'approver')
      .where('requester.id IN (:...employeeIds)', { employeeIds })
      .orderBy('request.createdAt', 'DESC')
      .getMany();
  }

  async findProcurementPending(): Promise<EquipmentRequest[]> {
    return this.requestRepository.find({
      where: { status: RequestStatus.PENDING_PROCUREMENT_APPROVAL },
      relations: {
        requester: { department: true },
        equipmentModel: { category: true },
        category: true,
        approvalSteps: { approver: true },
      },
      order: { createdAt: 'ASC', approvalSteps: { level: 'ASC' } },
    });
  }
}

File: src/EquipmentApi/Services/RequestService.cs
using EquipmentApi.Auth;
using EquipmentApi.Data;
using EquipmentApi.Dtos;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using EquipmentApi.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class RequestService
{
    private readonly AppDbContext _db;
    private readonly NotificationService _notificationService;
    private readonly AccessControlService _accessControl;
    private readonly ApprovalWorkflowService _approvalWorkflowService;

    public RequestService(
        AppDbContext db,
        NotificationService notificationService,
        AccessControlService accessControl,
        ApprovalWorkflowService approvalWorkflowService)
    {
        _db = db;
        _notificationService = notificationService;
        _accessControl = accessControl;
        _approvalWorkflowService = approvalWorkflowService;
    }

    public Task<List<EquipmentRequest>> FindMyRequestsAsync(Guid userId, CancellationToken cancellationToken = default)
    {
        return _db.EquipmentRequests
            .Where(request => request.RequesterId == userId)
            .Include(request => request.Requester)
            .Include(request => request.EquipmentModel!)
                .ThenInclude(model => model.Category)
            .Include(request => request.Category)
            .Include(request => request.ApprovalSteps)
                .ThenInclude(step => step.Approver)
            .OrderByDescending(request => request.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<EquipmentRequest> FindOneAsync(
        Guid id,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var request = await LoadRequestAsync(id, cancellationToken);
        if (request is null)
        {
            throw new NotFoundException($"Request with id \"{id}\" not found");
        }

        await AssertCanViewRequestAsync(request, user, cancellationToken);
        return request;
    }

    public async Task<object> GetTimelineAsync(
        Guid id,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var request = await FindOneAsync(id, user, cancellationToken);
        return new
        {
            requestId = request.Id,
            status = request.Status,
            steps = request.ApprovalSteps
                .OrderBy(step => step.Level)
                .Select(step => new
                {
                    id = step.Id,
                    level = step.Level,
                    approverRole = step.ApproverRole,
                    approverName = $"{step.Approver.FirstName} {step.Approver.LastName}",
                    status = step.Status,
                    comment = step.Comment,
                    actedAt = step.ActedAt,
                }),
            assignments = request.Assignments,
        };
    }

    public async Task<EquipmentRequest> CreateAsync(
        CreateRequestDto createRequestDto,
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var requester = await LoadRequesterAsync(user.Id, cancellationToken);
        AssertRequesterHasDirectManager(requester);

        EquipmentModel? equipmentModel = null;
        EquipmentCategory? category = null;

        if (createRequestDto.RequestType == RequestType.Loan)
        {
            var validated = await ValidateLoanRequestAsync(createRequestDto, cancellationToken);
            equipmentModel = validated.EquipmentModel;
            category = validated.Category;
        }
        else
        {
            category = await ValidateProcurementRequestAsync(createRequestDto, cancellationToken);
        }

        var savedRequest = await PersistNewRequestAsync(
            createRequestDto,
            requester,
            equipmentModel,
            category,
            cancellationToken);

        await NotifyInitialApproverAsync(savedRequest, cancellationToken);
        return savedRequest;
    }

    public async Task<EquipmentRequest> CancelAsync(
        Guid id,
        AuthenticatedUser user,
        CancelRequestDto cancelDto,
        CancellationToken cancellationToken = default)
    {
        var request = await FindOneAsync(id, user, cancellationToken);

        if (request.RequesterId != user.Id)
        {
            throw new ForbiddenException("You can only cancel your own requests");
        }

        if (!RequestStatusExtensions.CancellableStatuses.Contains(request.Status))
        {
            throw new BadRequestException(
                "Request can only be cancelled while pending manager or procurement approval");
        }

        request.Status = RequestStatus.Cancelled;
        request.CancellationReason = cancelDto.Reason;
        request.CancelledAt = DateTime.UtcNow;

        await _db.SaveChangesAsync(cancellationToken);
        await _notificationService.NotifyRequestCancelledAsync(request.Requester, request, cancellationToken);
        return request;
    }

    public async Task<List<EquipmentRequest>> FindManagerPendingAsync(
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var departmentIds = await _accessControl.GetManagedDepartmentIdsAsync(user.Id, cancellationToken);
        if (departmentIds.Count == 0)
        {
            return [];
        }

        var employeeIds = await _accessControl.GetDepartmentEmployeeIdsAsync(departmentIds, cancellationToken);
        if (employeeIds.Count == 0)
        {
            return [];
        }

        return await _db.EquipmentRequests
            .Where(request => request.Status == RequestStatus.PendingManagerApproval
                && employeeIds.Contains(request.RequesterId))
            .Include(request => request.Requester)
            .Include(request => request.EquipmentModel)
            .Include(request => request.Category)
            .Include(request => request.ApprovalSteps)
                .ThenInclude(step => step.Approver)
            .OrderByDescending(request => request.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<List<EquipmentRequest>> FindManagerRequestsAsync(
        AuthenticatedUser user,
        CancellationToken cancellationToken = default)
    {
        var departmentIds = await _accessControl.GetManagedDepartmentIdsAsync(user.Id, cancellationToken);
        if (departmentIds.Count == 0)
        {
            return [];
        }

        var employeeIds = await _accessControl.GetDepartmentEmployeeIdsAsync(departmentIds, cancellationToken);
        if (employeeIds.Count == 0)
        {
            return [];
        }

        return await _db.EquipmentRequests
            .Where(request => employeeIds.Contains(request.RequesterId))
            .Include(request => request.Requester)
            .Include(request => request.EquipmentModel)
            .Include(request => request.Category)
            .Include(request => request.ApprovalSteps)
                .ThenInclude(step => step.Approver)
            .OrderByDescending(request => request.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    public Task<List<EquipmentRequest>> FindProcurementPendingAsync(CancellationToken cancellationToken = default)
    {
        return _db.EquipmentRequests
            .Where(request => request.Status == RequestStatus.PendingProcurementApproval)
            .Include(request => request.Requester)
                .ThenInclude(requester => requester.Department)
            .Include(request => request.EquipmentModel!)
                .ThenInclude(model => model.Category)
            .Include(request => request.Category)
            .Include(request => request.ApprovalSteps)
                .ThenInclude(step => step.Approver)
            .OrderBy(request => request.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    private async Task<Employee> LoadRequesterAsync(Guid userId, CancellationToken cancellationToken)
    {
        var requester = await _db.Employees
            .Include(e => e.Department!)
                .ThenInclude(d => d.DirectManager)
            .FirstOrDefaultAsync(e => e.Id == userId, cancellationToken);

        if (requester is null)
        {
            throw new NotFoundException("Requester not found");
        }

        return requester;
    }

    private static void AssertRequesterHasDirectManager(Employee requester)
    {
        if (requester.Department?.DirectManager is null)
        {
            throw new BadRequestException(
                "Cannot submit request: department has no direct manager assigned");
        }
    }

    private async Task<(EquipmentModel EquipmentModel, EquipmentCategory Category)> ValidateLoanRequestAsync(
        CreateRequestDto createRequestDto,
        CancellationToken cancellationToken)
    {
        if (createRequestDto.EquipmentModelId is null)
        {
            throw new BadRequestException("equipmentModelId is required for loan requests");
        }

        var equipmentModel = await _db.EquipmentModels
            .Include(model => model.Category)
            .FirstOrDefaultAsync(model => model.Id == createRequestDto.EquipmentModelId, cancellationToken);

        if (equipmentModel is null)
        {
            throw new NotFoundException("Equipment model not found");
        }

        return (equipmentModel, equipmentModel.Category);
    }

    private async Task<EquipmentCategory> ValidateProcurementRequestAsync(
        CreateRequestDto createRequestDto,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(createRequestDto.RequestedItemName) || createRequestDto.CategoryId is null)
        {
            throw new BadRequestException(
                "requestedItemName and categoryId are required for procurement requests");
        }

        var category = await _db.EquipmentCategories
            .FirstOrDefaultAsync(c => c.Id == createRequestDto.CategoryId, cancellationToken);

        if (category is null)
        {
            throw new NotFoundException("Category not found");
        }

        return category;
    }

    private async Task<EquipmentRequest> PersistNewRequestAsync(
        CreateRequestDto createRequestDto,
        Employee requester,
        EquipmentModel? equipmentModel,
        EquipmentCategory? category,
        CancellationToken cancellationToken)
    {
        await using var transaction = await _db.Database.BeginTransactionAsync(cancellationToken);
        try
        {
            var request = new EquipmentRequest
            {
                RequesterId = requester.Id,
                RequestType = createRequestDto.RequestType,
                EquipmentModelId = equipmentModel?.Id,
                RequestedItemName = createRequestDto.RequestedItemName,
                CategoryId = category?.Id,
                Quantity = createRequestDto.Quantity ?? 1,
                StartDate = DateOnly.Parse(createRequestDto.StartDate),
                EndDate = DateOnly.Parse(createRequestDto.EndDate),
                Purpose = createRequestDto.Purpose,
                Status = RequestStatus.PendingManagerApproval,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow,
            };

            _db.EquipmentRequests.Add(request);
            await _db.SaveChangesAsync(cancellationToken);

            await _approvalWorkflowService.CreateInitialManagerApprovalStepAsync(
                _db,
                request,
                requester.Department!.DirectManager!,
                cancellationToken);

            await transaction.CommitAsync(cancellationToken);

            return await LoadRequestAsync(request.Id, cancellationToken)
                ?? throw new NotFoundException($"Request with id \"{request.Id}\" not found");
        }
        catch
        {
            await transaction.RollbackAsync(cancellationToken);
            throw;
        }
    }

    private async Task NotifyInitialApproverAsync(
        EquipmentRequest savedRequest,
        CancellationToken cancellationToken)
    {
        var firstStep = savedRequest.ApprovalSteps.OrderBy(step => step.Level).FirstOrDefault();
        if (firstStep?.Approver is not null)
        {
            await _notificationService.NotifyApprovalRequiredAsync(
                firstStep.Approver,
                savedRequest,
                firstStep,
                cancellationToken);
        }
    }

    private async Task AssertCanViewRequestAsync(
        EquipmentRequest request,
        AuthenticatedUser user,
        CancellationToken cancellationToken)
    {
        if (request.RequesterId == user.Id)
        {
            return;
        }

        if (_accessControl.IsAdmin(user) || _accessControl.IsProcurementManagerOrAbove(user))
        {
            return;
        }

        if (_accessControl.IsDirectManagerOrAbove(user))
        {
            var canView = await _accessControl.IsEmployeeInManagedDepartmentsAsync(
                user.Id,
                request.RequesterId,
                cancellationToken);
            if (canView)
            {
                return;
            }
        }

        throw new ForbiddenException("You cannot view this request");
    }

    private Task<EquipmentRequest?> LoadRequestAsync(Guid id, CancellationToken cancellationToken)
    {
        return _db.EquipmentRequests
            .Include(request => request.Requester)
                .ThenInclude(requester => requester.Department)
            .Include(request => request.EquipmentModel!)
                .ThenInclude(model => model.Category)
            .Include(request => request.Category)
            .Include(request => request.ApprovalSteps)
                .ThenInclude(step => step.Approver)
            .Include(request => request.Assignments)
                .ThenInclude(assignment => assignment.EquipmentAsset)
                    .ThenInclude(asset => asset.EquipmentModel)
            .FirstOrDefaultAsync(request => request.Id == id, cancellationToken);
    }
}
